### ISSUE TITLE
union type implementing exclusive or required parameter does not work as function parameter type

### ISSUE THREAD

**Issue by gr2m** [Microsoft/TypeScript#30800]
**TypeScript Version:**  3.5.0-dev.20190406

**Search Terms:** union required properties

**Code**

```ts
declare interface Base {
  optional?: string
}

declare interface Options1 extends Base {
  required1: string
}

declare interface Options2 extends Base {
  required2: string
}

export type Options = Options1 | Options2;

function test (options: Options) {
  console.log(options.optional)
  console.log(options.required1)
  console.log(options.required2)
}

test({
  optional: 'yo',
  required1: '123'
})
```

**Expected behavior:**

It should not complain about `options.required1` or `options.required2`

**Actual behavior:**

I get an Error:

> Property 'required1' does not exist on type 'Options'

**Playground Link:** https://www.typescriptlang.org/play/#src=declare%20interface%20Base%20%7B%0D%0A%20%20optional%3F%3A%20string%0D%0A%7D%0D%0A%0D%0Adeclare%20interface%20Options1%20extends%20Base%20%7B%0D%0A%20%20required1%3A%20string%0D%0A%7D%0D%0A%0D%0Adeclare%20interface%20Options2%20extends%20Base%20%7B%0D%0A%20%20required2%3A%20string%0D%0A%7D%0D%0A%0D%0Aexport%20type%20Options%20%3D%20Options1%20%7C%20Options2%3B%0D%0A%0D%0Afunction%20test%20(options%3A%20Options)%20%7B%0D%0A%20%20console.log(options.optional)%0D%0A%20%20console.log(options.required1)%0D%0A%20%20console.log(options.required2)%0D%0A%7D%0D%0A%0D%0Atest(%7B%0D%0A%20%20optional%3A%20'yo'%2C%0D%0A%20%20required1%3A%20'123'%0D%0A%7D)

**Related Issues:** I only found this StackOverflow question: https://stackoverflow.com/questions/40510611/typescript-interface-require-one-of-two-properties-to-exist. The solution works until you want to use the defined type as a function parameter

**Comment by RyanCavanaugh** [comment]
TypeScript intentionally does not allow access to possibly-unpresent properties without doing some check first to see if they exist

**Comment by gr2m** [comment]
Thank you Ryan, checking explicitly if the property exists fixed the problem

```js
  if ('required1' in options) {
    console.log(options.required1)  
  }
```
